File: PersonalizedMemoryCalibration.js
// ============================================================================
// Personalized Memory Calibration - Bayesian Parameter Adjustment
// Learns from individual practice patterns to personalize τ-parameters
// Copyright © 2025 Frank De Baere - All Rights Reserved
// ============================================================================

/**
 * PersonalizedMemoryCalibration - Learns individual forgetting patterns
 * Uses Bayesian learning to adapt tau parameters based on actual performance
 */
class PersonalizedMemoryCalibration {
    constructor(storagePrefix = 'mp_') {
        this.storagePrefix = storagePrefix;
        this.calibrationData = null;
        this.profileName = null;
    }

    /**
     * Initializes calibration system for a specific profile
     */
    initializeCalibrationSystem(profileName) {
        try {
            this.profileName = profileName;
            this.loadCalibrationData();
            
            console.log(`[PMC] Initialized calibration system for profile: ${profileName}`);
        } catch (error) {
            console.error('[PMC] Failed to initialize calibration system:', error);
            this.calibrationData = this.createEmptyCalibrationData();
        }
    }

    /**
     * Calculates personalized τ-value based on individual history
     */
    getPersonalizedTau(difficulty, repetitions) {
        // Start with scientific basis
        const baseTau = EbbinghausConstants.calculateAdjustedTau(difficulty, repetitions);

        // Apply personal adjustment based on calibration
        const personalAdjustment = this.calculatePersonalAdjustment(difficulty, repetitions);

        const personalizedTau = baseTau * personalAdjustment;

        console.log(`[PMC] Personal Tau Calculation: Base Tau: ${baseTau.toFixed(2)} days, Personal Adjustment: ${personalAdjustment.toFixed(2)}x, Personalized Tau: ${personalizedTau.toFixed(2)} days`);

        return personalizedTau;
    }

    /**
     * Updates calibration from a practice session
     */
    updateCalibrationFromSession(session, section) {
        if (!session) {
            console.warn('[PMC] Session is null; skipping calibration update');
            return;
        }

        // Determine difficulty (fallback to Average)
        let difficulty = 'Average';
        try {
            if (section && section.difficulty) {
                difficulty = section.difficulty;
            }
        } catch (error) {
            // Ignore - difficulty stays "Average"
        }

        try {
            // 1. Calculate prediction accuracy
            const predictionAccuracy = section ? 
                this.calculatePredictionAccuracy(session, section) : 1.0;

            // 2. Update personal parameters for this difficulty
            this.updatePersonalParameters(difficulty, predictionAccuracy);

            // 3. Persist changes
            this.saveCalibrationData();

            console.log(
                `[PMC] Updated calibration. Diff=${difficulty}, ` +
                `Outcome='${session.sessionOutcome}', Reps=${session.repetitions}, ` +
                `PredAccuracy=${predictionAccuracy.toFixed(3)}`
            );
        } catch (error) {
            console.error('[PMC] Error during calibration update:', error);
        }
    }

    /**
     * Gets calibration statistics for debugging/UI
     */
    getCalibrationStats() {
        if (!this.calibrationData) {
            return {
                totalSessions: 0,
                isCalibrated: false,
                difficultyAdjustments: {}
            };
        }

        return {
            totalSessions: this.calibrationData.totalSessions,
            isCalibrated: this.hasSufficientDataForCalibration(),
            difficultyAdjustments: Object.fromEntries(
                Object.entries(this.calibrationData.difficultyAdjustments).map(([key, value]) => [
                    key,
                    {
                        factor: value.adjustmentFactor,
                        confidence: value.confidence,
                        sessions: value.sessionCount
                    }
                ])
            )
        };
    }

    // ========================================================================
    // PRIVATE HELPER METHODS
    // ========================================================================

    /**
     * Calculates personal adjustment factor based on collected data
     * Supports rapid calibration with lower session requirements
     */
    calculatePersonalAdjustment(difficulty, repetitions) {
        // Check for rapid calibration phase (first 5 sessions get immediate but conservative adjustments)
        const isRapidPhase = this.calibrationData.totalSessions <= 5;

        if (!this.hasSufficientDataForCalibration() && !isRapidPhase) {
            return 1.0; // Use standard values if insufficient data
        }

        // Find relevant calibration data for this difficulty
        const difficultyKey = difficulty.toLowerCase();
        const adjustment = this.calibrationData.difficultyAdjustments[difficultyKey];

        if (!adjustment) {
            return 1.0;
        }

        let confidenceFactor;

        if (isRapidPhase) {
            // Rapid calibration: lower confidence but still apply adjustments
            confidenceFactor = Math.min(0.6, adjustment.sessionCount / 3.0); // Max 60% confidence in rapid phase
            console.log(`[PMC] Rapid calibration active for ${difficulty} (sessions=${adjustment.sessionCount}, confidence=${confidenceFactor.toFixed(2)})`);
        } else {
            // Normal calibration: gradual confidence buildup
            confidenceFactor = Math.min(1.0, this.calibrationData.totalSessions / 25.0); // Reduced from 50 to 25
        }

        const personalizedFactor = 1.0 + (adjustment.adjustmentFactor - 1.0) * confidenceFactor;

        return Math.max(0.3, Math.min(3.0, personalizedFactor)); // Safe bounds
    }

    /**
     * Calculates how accurate our interval prediction was
     */
    calculatePredictionAccuracy(session, section) {
        // Calculate expected retention based on interval
        const lastPracticeDate = section.lastPracticeDate ? new Date(section.lastPracticeDate) : new Date(session.date);
        const sessionDate = new Date(session.date);
        const daysSincePractice = (sessionDate - lastPracticeDate) / (1000 * 60 * 60 * 24);

        if (daysSincePractice <= 0) return 1.0;

        const expectedTau = EbbinghausConstants.calculateAdjustedTau(
            section.difficulty,
            section.completedRepetitions || 0
        );
        const expectedRetention = EbbinghausConstants.calculateRetention(
            daysSincePractice,
            expectedTau
        );

        // Estimate actual retention based on session performance
        const actualRetention = this.estimateActualRetention(session);

        // Calculate accuracy (1.0 = perfect, 0.0 = completely wrong)
        const accuracy = 1.0 - Math.abs(expectedRetention - actualRetention);
        return Math.max(0.0, Math.min(1.0, accuracy));
    }

    /**
     * Estimates actual retention based on session performance
     * Uses multiple factors for accurate estimation
     */
    estimateActualRetention(session) {
        // Use performanceScore if available (most reliable)
        if (session.performanceScore !== undefined && session.performanceScore !== null) {
            // Map performance score (0-10) to retention (0-1)
            // Poor (2.5) → 0.3, Fair (5.0) → 0.6, Good (7.5) → 0.8, Excellent (9.5) → 0.95
            const retention = 0.2 + (session.performanceScore / 10.0) * 0.75;
            return Math.max(0.1, Math.min(1.0, retention));
        }

        // Fallback: estimate based on efficiency metrics
        if (!session.duration || session.duration <= 0) return 0.5;

        const durationMinutes = session.duration / (1000 * 60); // Convert ms to minutes
        
        // Avoid division by zero
        if (durationMinutes <= 0) return 0.5;
        
        const repetitionsPerMinute = session.repetitions / durationMinutes;

        // Normalize to 0-1 scale (0.5-2.0 rep/min = normal range)
        let normalizedEfficiency = Math.max(0.1, Math.min(1.0, repetitionsPerMinute / 2.0));

        // Adjustment based on attempts till success (execution failures)
        const executionFails = getExecutionFailures(session);
        if (executionFails && executionFails > 0) {
            const attemptsAdjustment = 1.0 / Math.sqrt(executionFails);
            normalizedEfficiency *= attemptsAdjustment;
        }

        // Adjustment based on memory failures (streak resets)
        const memoryFails = getMemoryFailures(session);
        if (memoryFails && memoryFails > 0) {
            const failureRatio = memoryFails / (session.repetitions + memoryFails);
            normalizedEfficiency *= (1.0 - failureRatio * 0.5); // Failures reduce efficiency
        }

        return Math.max(0.1, Math.min(1.0, normalizedEfficiency));
    }

    /**
     * Updates personal parameters with Bayesian learning
     */
    updatePersonalParameters(difficulty, accuracy) {
        const difficultyKey = difficulty.toLowerCase();

        if (!this.calibrationData.difficultyAdjustments[difficultyKey]) {
            this.calibrationData.difficultyAdjustments[difficultyKey] = {
                adjustmentFactor: 1.0,
                confidence: 0.0,
                sessionCount: 0
            };
        }

        const adjustment = this.calibrationData.difficultyAdjustments[difficultyKey];

        // Bayesian update with learning rate
        const learningRate = EbbinghausConstants.BAYESIAN_LEARNING_RATE;
        
        let targetAdjustment;
        if (accuracy < 0.5) {
            targetAdjustment = accuracy < 0.3 ? 0.7 : 0.85; // Too fast forgetting -> shorter tau
        } else {
            targetAdjustment = accuracy > 0.8 ? 1.3 : 1.15; // Too good retention -> longer tau
        }

        // Gradual adjustment towards target
        adjustment.adjustmentFactor += learningRate * (targetAdjustment - adjustment.adjustmentFactor);
        adjustment.sessionCount++;
        adjustment.confidence = Math.min(1.0, adjustment.sessionCount / 20.0);

        this.calibrationData.totalSessions++;
    }

    /**
     * Checks if there is sufficient data for reliable calibration
     * Reduced minimum sessions for faster adaptation
     */
    hasSufficientDataForCalibration() {
        const ACCELERATED_MIN_SESSIONS = 5; // Reduced from 10 to 5
        return this.calibrationData.totalSessions >= ACCELERATED_MIN_SESSIONS;
    }

    /**
     * Loads calibration data from localStorage
     */
    loadCalibrationData() {
        try {
            if (!this.profileName) {
                this.calibrationData = this.createEmptyCalibrationData();
                return;
            }

            const key = `${this.storagePrefix}${this.profileName}_calibration`;
            const jsonContent = localStorage.getItem(key);

            if (jsonContent) {
                this.calibrationData = JSON.parse(jsonContent);
                console.log(`[PMC] Loaded calibration data for profile: ${this.profileName}`);
            } else {
                this.calibrationData = this.createEmptyCalibrationData();
                console.log(`[PMC] No existing calibration data found, initialized new data`);
            }
        } catch (error) {
            console.error('[PMC] Error loading calibration data:', error);
            this.calibrationData = this.createEmptyCalibrationData();
        }
    }

    /**
     * Saves calibration data to localStorage
     */
    saveCalibrationData() {
        try {
            if (!this.profileName || !this.calibrationData) {
                console.warn('[PMC] Cannot save calibration data: missing profile or data');
                return;
            }

            const key = `${this.storagePrefix}${this.profileName}_calibration`;
            this.calibrationData.lastCalibrationUpdate = new Date().toISOString();
            
            const jsonContent = JSON.stringify(this.calibrationData);
            localStorage.setItem(key, jsonContent);

            console.log(`[PMC] Saved calibration data for profile: ${this.profileName}`);
        } catch (error) {
            console.error('[PMC] Error saving calibration data:', error);
        }
    }

    /**
     * Creates empty calibration data structure
     */
    createEmptyCalibrationData() {
        return {
            totalSessions: 0,
            difficultyAdjustments: {
                // Will be populated as sessions are recorded
                // Keys: 'difficult', 'average', 'easy', 'mastered'
                // Values: { adjustmentFactor, confidence, sessionCount }
            },
            lastCalibrationUpdate: new Date().toISOString()
        };
    }

    /**
     * Resets calibration data (useful for debugging or user request)
     */
    resetCalibration() {
        this.calibrationData = this.createEmptyCalibrationData();
        this.saveCalibrationData();
        console.log('[PMC] Calibration data reset');
    }

    /**
     * Gets detailed calibration report for diagnostics
     */
    getDetailedCalibrationReport() {
        if (!this.calibrationData) {
            return 'No calibration data available';
        }

        const difficulties = ['difficult', 'average', 'easy', 'mastered'];
        let report = `\n${'='.repeat(60)}\n`;
        report += `📊 PERSONALIZED MEMORY CALIBRATION REPORT\n`;
        report += `${'='.repeat(60)}\n`;
        report += `Profile: ${this.profileName || 'Unknown'}\n`;
        report += `Total Sessions: ${this.calibrationData.totalSessions}\n`;
        report += `Calibration Status: ${this.hasSufficientDataForCalibration() ? '✅ Active' : '⏳ Learning'}\n`;
        report += `Last Update: ${new Date(this.calibrationData.lastCalibrationUpdate).toLocaleString()}\n`;
        report += `${'='.repeat(60)}\n`;

        difficulties.forEach(diff => {
            const adj = this.calibrationData.difficultyAdjustments[diff];
            if (adj && adj.sessionCount > 0) {
                report += `\n${diff.toUpperCase()}:\n`;
                report += `  Adjustment Factor: ${adj.adjustmentFactor.toFixed(3)}x\n`;
                report += `  Confidence: ${(adj.confidence * 100).toFixed(1)}%\n`;
                report += `  Sessions: ${adj.sessionCount}\n`;
                
                // Interpret adjustment
                let interpretation = '';
                if (adj.adjustmentFactor < 0.9) {
                    interpretation = '⚡ Faster forgetting → Shorter intervals';
                } else if (adj.adjustmentFactor > 1.1) {
                    interpretation = '🧠 Stronger memory → Longer intervals';
                } else {
                    interpretation = '✓ Standard retention';
                }
                report += `  Interpretation: ${interpretation}\n`;
            }
        });

        report += `\n${'='.repeat(60)}\n`;
        return report;
    }

    /**
     * Exports calibration data for backup/analysis
     */
    exportCalibrationData() {
        if (!this.calibrationData) {
            return null;
        }

        return {
            version: '1.0',
            exportDate: new Date().toISOString(),
            profileName: this.profileName,
            calibrationData: JSON.parse(JSON.stringify(this.calibrationData))
        };
    }

    /**
     * Imports calibration data from backup
     */
    importCalibrationData(exportedData) {
        try {
            if (!exportedData || !exportedData.calibrationData) {
                throw new Error('Invalid calibration data format');
            }

            this.calibrationData = exportedData.calibrationData;
            this.saveCalibrationData();
            
            console.log('[PMC] Successfully imported calibration data');
            return true;
        } catch (error) {
            console.error('[PMC] Error importing calibration data:', error);
            return false;
        }
    }
}

// Export for use in other modules
if (typeof module !== 'undefined' && module.exports) {
    module.exports = { PersonalizedMemoryCalibration };
}

File: 01ReferentieBronCode/PersonalizedMemoryCalibration.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;

namespace ModusPractica
{
    /// <summary>
    /// Persoonlijke geheugen kalibratie gebaseerd op Bayesiaanse parameter aanpassing
    /// Leert van individuele oefenpatronen om τ-parameters te personaliseren
    /// </summary>
    public class PersonalizedMemoryCalibration
    {
        private static PersonalizedMemoryCalibration _instance;
        private PersonalCalibrationData _calibrationData;
        private string _calibrationFilePath;

        // Singleton pattern
        public static PersonalizedMemoryCalibration Instance
        {
            get
            {
                if (_instance == null)
                {
                    _instance = new PersonalizedMemoryCalibration();
                }
                return _instance;
            }
        }

        private PersonalizedMemoryCalibration()
        {
            InitializeCalibrationSystem();
        }

        public void InitializeCalibrationSystem()
        {
            if (!RetentionFeatureFlags.UsePMC)
            {
                // Disabled: do not initialize or touch the filesystem
                return;
            }
            try
            {
                // Profiel: gebruik het actieve profiel en centrale padprovider
                string calibrationFolder = Path.Combine(
                    DataPathProvider.GetProfileFolder(ActiveUserSession.ProfileName),
                    "Calibration");

                if (!Directory.Exists(calibrationFolder))
                {
                    Directory.CreateDirectory(calibrationFolder);
                }

                _calibrationFilePath = Path.Combine(
                    calibrationFolder,
                    "personal_memory_calibration.json"
                );

                // Handige logging: waar wordt PMC opgeslagen/gelezen?
                MLLogManager.Instance?.Log(
                    $"PMC.InitializeCalibrationSystem: Calibration file path set to '{_calibrationFilePath}'.",
                    LogLevel.Info
                );

                // Laad bestaande kalibratie of initialiseer nieuwe
                LoadCalibrationData();
            }
            catch (Exception ex)
            {
                MLLogManager.Instance?.LogError("PMC.InitializeCalibrationSystem: Failed to initialize calibration system.", ex);
                // Fail closed: laat _calibrationFilePath in de laatste bekende staat;
                // verdere calls (Save/Load) loggen hun eigen fouten.
            }
        }



        /// <summary>
        /// Berekent gepersonaliseerde τ-waarde op basis van individuele geschiedenis
        /// </summary>
        public double GetPersonalizedTau(string difficulty, int repetitions)
        {
            // If disabled, return standard calculation
            if (!RetentionFeatureFlags.UsePMC)
            {
                return EbbinghausConstants.CalculateAdjustedTau(difficulty, repetitions);
            }
            // Start met wetenschappelijke basis
            double baseTau = EbbinghausConstants.CalculateAdjustedTau(difficulty, repetitions);

            // Pas aan op basis van persoonlijke kalibratie
            double personalAdjustment = CalculatePersonalAdjustment(difficulty, repetitions);

            double personalizedTau = baseTau * personalAdjustment;

            // Log voor debugging
            MLLogManager.Instance.Log($"Personal Tau Calculation: Base Tau: {baseTau:F2} days, Personal Adjustment: {personalAdjustment:F2}x, Personalized Tau: {personalizedTau:F2} days", LogLevel.Info);

            return personalizedTau;
        }

        public void UpdateCalibrationFromSession(PracticeHistory session, BarSection section)
        {
            if (!RetentionFeatureFlags.UsePMC)
            {
                return; // disabled
            }
            if (session == null)
            {
                MLLogManager.Instance?.Log("PMC.UpdateCalibrationFromSession: session is null; skipping.", LogLevel.Warning);
                return;
            }

            // Difficulty bepalen (fallback naar Average)
            string difficulty = "Average";
            try
            {
                if (section != null && !string.IsNullOrWhiteSpace(section.Difficulty))
                    difficulty = section.Difficulty;
            }
            catch
            {
                // ignore – difficulty blijft "Average"
            }

            try
            {
                // 1) Bereken voorspelfout/accuracy (interne methode gebruikt o.a. τ-verwachting)
                double predictionAccuracy = section != null ? CalculatePredictionAccuracy(session, section) : 1.0;

                // 2) Pas persoonlijke parameters aan voor de betreffende moeilijkheid
                UpdatePersonalParameters(difficulty, predictionAccuracy);

                // 3) Persist + logging
                SaveCalibrationData();

                MLLogManager.Instance?.Log(
                    $"PMC.UpdateCalibrationFromSession: Updated calibration. Diff={difficulty}, " +
                    $"Outcome='{session.SessionOutcome}', Reps={session.Repetitions}, " +
                    $"PredAccuracy={predictionAccuracy:F3}",
                    LogLevel.Info
                );
            }
            catch (Exception ex)
            {
                MLLogManager.Instance?.LogError("PMC.UpdateCalibrationFromSession: Error during calibration update.", ex);
            }
        }



        /// <summary>
        /// Berekent persoonlijke aanpassingsfactor op basis van opgebouwde data
        /// UPDATED: Supports rapid calibration with lower session requirements
        /// </summary>
        private double CalculatePersonalAdjustment(string difficulty, int repetitions)
        {
            // NEW: Check for rapid calibration phase (first 5 sessions get immediate but conservative adjustments)
            bool isRapidPhase = _calibrationData.TotalSessions <= 5;

            if (!HasSufficientDataForCalibration() && !isRapidPhase)
            {
                return 1.0; // Gebruik standaard waarden als onvoldoende data
            }

            // Zoek relevante kalibratie data voor deze moeilijkheidsgraad
            if (_calibrationData.DifficultyAdjustments.TryGetValue(difficulty.ToLower(), out var adjustment))
            {
                double confidenceFactor;

                if (isRapidPhase)
                {
                    // Rapid calibration: lower confidence but still apply adjustments
                    confidenceFactor = Math.Min(0.6, adjustment.SessionCount / 3.0); // Max 60% confidence in rapid phase
                    MLLogManager.Instance?.Log(
                        $"PMC: Rapid calibration active for {difficulty} (sessions={adjustment.SessionCount}, confidence={confidenceFactor:F2})",
                        LogLevel.Debug);
                }
                else
                {
                    // Normal calibration: gradual confidence buildup
                    confidenceFactor = Math.Min(1.0, _calibrationData.TotalSessions / 25.0); // Reduced from 50 to 25
                }

                double personalizedFactor = 1.0 + (adjustment.AdjustmentFactor - 1.0) * confidenceFactor;

                return Math.Max(0.3, Math.Min(3.0, personalizedFactor)); // Veilige grenzen
            }

            return 1.0;
        }

        /// <summary>
        /// Berekent hoe accuraat onze interval voorspelling was
        /// </summary>
        private double CalculatePredictionAccuracy(PracticeHistory session, BarSection section)
        {
            // Bereken verwachte retentie op basis van interval
            double daysSincePractice = (session.Date - (section.LastPracticeDate ?? session.Date)).TotalDays;
            if (daysSincePractice <= 0) return 1.0;

            double expectedTau = EbbinghausConstants.CalculateAdjustedTau(section.Difficulty, section.CompletedRepetitions);
            double expectedRetention = EbbinghausConstants.CalculateRetention(daysSincePractice, expectedTau);

            // Schat werkelijke retentie op basis van sessie prestatie
            double actualRetention = EstimateActualRetention(session);

            // Bereken accuraatheid (1.0 = perfect, 0.0 = volledig verkeerd)
            double accuracy = 1.0 - Math.Abs(expectedRetention - actualRetention);
            return Math.Max(0.0, Math.Min(1.0, accuracy));
        }

        /// <summary>
        /// Schat werkelijke retentie op basis van sessie prestaties
        /// </summary>
        private double EstimateActualRetention(PracticeHistory session)
        {
            // Eenvoudige schatting op basis van herhalingen en tijd
            // Meer herhalingen in minder tijd = betere retentie
            if (session.Duration.TotalMinutes <= 0) return 0.5;

            double repetitionsPerMinute = session.Repetitions / session.Duration.TotalMinutes;

            // Normaliseer naar 0-1 schaal (0.5-2.0 rep/min = normale range)
            double normalizedEfficiency = Math.Max(0.1, Math.Min(1.0, repetitionsPerMinute / 2.0));

            // Aanpassing op basis van attempts till success
            if (session.AttemptsTillSuccess > 0)
            {
                double attemptsAdjustment = 1.0 / Math.Sqrt(session.AttemptsTillSuccess);
                normalizedEfficiency *= attemptsAdjustment;
            }

            return Math.Max(0.1, Math.Min(1.0, normalizedEfficiency));
        }

        /// <summary>
        /// Update persoonlijke parameters met Bayesiaanse learning
        /// </summary>
        private void UpdatePersonalParameters(string difficulty, double accuracy)
        {
            string difficultyKey = difficulty.ToLower();

            if (!_calibrationData.DifficultyAdjustments.ContainsKey(difficultyKey))
            {
                _calibrationData.DifficultyAdjustments[difficultyKey] = new DifficultyAdjustment
                {
                    AdjustmentFactor = 1.0,
                    Confidence = 0.0,
                    SessionCount = 0
                };
            }

            var adjustment = _calibrationData.DifficultyAdjustments[difficultyKey];

            // Bayesiaanse update met learning rate
            double learningRate = EbbinghausConstants.BAYESIAN_LEARNING_RATE;
            double targetAdjustment = accuracy < 0.5 ?
                (accuracy < 0.3 ? 0.7 : 0.85) :  // Te snel vergeten -> korter tau
                (accuracy > 0.8 ? 1.3 : 1.15);   // Te goed onthouden -> langer tau

            // Geleidelijke aanpassing naar doel
            adjustment.AdjustmentFactor += learningRate * (targetAdjustment - adjustment.AdjustmentFactor);
            adjustment.SessionCount++;
            adjustment.Confidence = Math.Min(1.0, adjustment.SessionCount / 20.0);

            _calibrationData.TotalSessions++;
        }

        /// <summary>
        /// Controleert of er voldoende data is voor betrouwbare kalibratie
        /// UPDATED: Reduced minimum sessions for faster adaptation
        /// </summary>
        private bool HasSufficientDataForCalibration()
        {
            // Reduced from 10 to 5 sessions for faster user adaptation
            const int ACCELERATED_MIN_SESSIONS = 5;
            return _calibrationData.TotalSessions >= ACCELERATED_MIN_SESSIONS;
        }

        /// <summary>
        /// Laadt kalibratie data uit bestand
        /// </summary>
        private void LoadCalibrationData()
        {
            if (!RetentionFeatureFlags.UsePMC)
            {
                _calibrationData = new PersonalCalibrationData();
                return;
            }
            try
            {
                if (File.Exists(_calibrationFilePath))
                {
                    string jsonContent = File.ReadAllText(_calibrationFilePath);
                    _calibrationData = JsonSerializer.Deserialize<PersonalCalibrationData>(jsonContent)
                                     ?? new PersonalCalibrationData();
                }
                else
                {
                    _calibrationData = new PersonalCalibrationData();
                }
            }
            catch (Exception ex)
            {
                MLLogManager.Instance.LogError($"Error loading calibration data from {_calibrationFilePath}: {ex.Message}", ex);
                _calibrationData = new PersonalCalibrationData();
            }
        }

        /// <summary>
        /// Slaat kalibratie data op naar bestand
        /// </summary>
        private void SaveCalibrationData()
        {
            if (!RetentionFeatureFlags.UsePMC) return;
            try
            {
                string jsonContent = JsonSerializer.Serialize(_calibrationData, new JsonSerializerOptions
                {
                    WriteIndented = true
                });

                File.WriteAllText(_calibrationFilePath, jsonContent);
            }
            catch (Exception ex)
            {
                MLLogManager.Instance.LogError($"Error saving calibration data to {_calibrationFilePath}: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Geeft calibratie statistieken voor debugging
        /// </summary>
        public CalibrationStats GetCalibrationStats()
        {
            if (!RetentionFeatureFlags.UsePMC)
            {
                return new CalibrationStats
                {
                    TotalSessions = 0,
                    IsCalibrated = false,
                    DifficultyAdjustments = new Dictionary<string, object>()
                };
            }
            return new CalibrationStats
            {
                TotalSessions = _calibrationData.TotalSessions,
                IsCalibrated = HasSufficientDataForCalibration(),
                DifficultyAdjustments = _calibrationData.DifficultyAdjustments.ToDictionary(
                    kvp => kvp.Key,
                    kvp => (object)new
                    {
                        Factor = kvp.Value.AdjustmentFactor,
                        Confidence = kvp.Value.Confidence,
                        Sessions = kvp.Value.SessionCount
                    })
            };
        }
    }

    #region Data Classes

    /// <summary>
    /// Data klasse voor persoonlijke kalibratie informatie
    /// </summary>
    public class PersonalCalibrationData
    {
        public int TotalSessions { get; set; } = 0;
        public Dictionary<string, DifficultyAdjustment> DifficultyAdjustments { get; set; } =
            new Dictionary<string, DifficultyAdjustment>();
        public DateTime LastCalibrationUpdate { get; set; } = DateTime.Now;
    }

    /// <summary>
    /// Aanpassingsdata per moeilijkheidsgraad
    /// </summary>
    public class DifficultyAdjustment
    {
        public double AdjustmentFactor { get; set; } = 1.0;
        public double Confidence { get; set; } = 0.0;
        public int SessionCount { get; set; } = 0;
    }

    /// <summary>
    /// Kalibratie statistieken voor UI weergave
    /// </summary>
    public class CalibrationStats
    {
        public int TotalSessions { get; set; }
        public bool IsCalibrated { get; set; }
        public Dictionary<string, object> DifficultyAdjustments { get; set; } = new Dictionary<string, object>();
    }

    #endregion Data Classes
}